File: services/extractor/src/rung2.ts
import { companyFromDisplayName, matchGenericBody } from '@loop/rules';
import { domainOfAddress, matchesDomainSuffix, type Intent, type RawMessage } from '@loop/domain';

/**
 * Rung 2 — calendar and thread heuristics.
 *
 * "An .ics attachment or a meeting link from a company domain resolves the
 * stage by itself. A reply on a known thread inherits the application with no
 * parsing at all." 13% of messages, 2 ms, €0.
 */

export interface Rung2Context {
  /** Threads already attached to an application, mapped to that application. */
  threadToApplication: ReadonlyMap<string, string>;
  /** Domains that belong to an ATS rather than to an employer. */
  atsDomains: readonly string[];
}

export interface Rung2Result {
  intent: Intent;
  confidence: number;
  stageHint: string | null;
  company: string | null;
  startsAt: string | null;
  endsAt: string | null;
  calendarEventId: string | null;
  applicationId: string | null;
  location: string | null;
}

/**
 * The stage a calendar title implies.
 *
 * Interview invitations name their own stage more often than any other signal
 * in the mailbox — "System & code review", "HR screening", "Final round" — and
 * a title match is both cheaper and more reliable than asking a model.
 */
const TITLE_TO_STAGE: Array<[RegExp, string]> = [
  [/\b(final|exec|leadership|founder|ceo|cto)\b/i, 'final'],
  [/\b(onsite|on-site|loop|super\s*day|assessment centre|full day)\b/i, 'onsite_loop'],
  [/\b(system\s*design|architecture|design (round|interview))\b/i, 'system_design'],
  [/\b(technical|coding|live\s*cod|pair\s*program|algorithm|code review)\b/i, 'technical'],
  [/\b(take[- ]home|assignment|exercise|challenge)\b/i, 'take_home'],
  [/\b(hr|people|talent|recruiter|screening|intro|introductory|first call|knowledge)\b/i, 'hr_call'],
];

export function stageFromTitle(title: string | null): string | null {
  if (!title) return null;
  for (const [re, stage] of TITLE_TO_STAGE) if (re.test(title)) return stage;
  return null;
}

export function runRung2(msg: RawMessage, ctx: Rung2Context): Rung2Result | null {
  const senderDomain = domainOfAddress(msg.headers.from);
  const isAts = !!senderDomain && ctx.atsDomains.some((d) => matchesDomainSuffix(senderDomain, d));

  // ── a calendar invite ────────────────────────────────────────────────────
  if (msg.invite) {
    const inv = msg.invite;
    const cancelled = inv.status === 'cancelled' || inv.method === 'CANCEL';
    const stage = stageFromTitle(inv.summary) ?? 'technical';
    return {
      intent: cancelled ? 'interview_cancelled' : 'interview_invite',
      // "Every .ics invite from a company domain is a near-certain interview."
      confidence: cancelled ? 0.95 : 0.97,
      stageHint: stage,
      company: companyFromOrganiser(inv.organiser, ctx.atsDomains),
      startsAt: inv.starts_at,
      endsAt: inv.ends_at,
      calendarEventId: inv.uid,
      applicationId: msg.thread_id ? (ctx.threadToApplication.get(msg.thread_id) ?? null) : null,
      location: inv.location,
    };
  }

  // ── a reply on a thread we already own ───────────────────────────────────
  // No parsing at all: the thread identity is the strongest and cheapest signal
  // in the system, so this abstains on *intent* and only asserts identity.
  if (msg.thread_id && ctx.threadToApplication.has(msg.thread_id) && !isAts) {
    return {
      intent: 'other',
      confidence: 0.99,
      stageHint: null,
      company: null,
      startsAt: null,
      endsAt: null,
      calendarEventId: null,
      applicationId: ctx.threadToApplication.get(msg.thread_id)!,
      location: null,
    };
  }

  // ── deterministic body vocabulary, from any sender ───────────────────────
  //
  // The ladder's rung 3 exists because "unknown template, human-written email,
  // Italian/English mixed prose" cannot be pattern-matched. That is true of the
  // *general* case — but not of the commonest sentences in recruiting, which
  // are close to formulaic in both languages: "non proseguiremo", "vorremmo
  // invitarti", "grazie per la tua candidatura".
  //
  // Reading those without a model matters here specifically: the box this runs
  // on is an 8 GB laptop, where a resident 7B alongside Postgres and eight
  // services is not a trade worth making. So the phrases a rule can honestly
  // recognise are recognised, at a confidence a step below a real ATS template,
  // and everything genuinely ambiguous still goes to a human.
  //
  // Only for mail the classifier already scored as job-related — this never
  // sees the general inbox.
  const generic = matchGenericBody(`${msg.headers.subject}\n${msg.text}`);
  if (generic && !isAts) {
    return {
      intent: generic.intent,
      // One step below the same phrase from a known ATS: the sender is not
      // established, so the claim is weaker even when the words are identical.
      confidence: Math.min(generic.confidence, 0.88),
      stageHint: null,
      company: companyFromDisplayName(msg.headers.from) ?? companyFromOrganiser(senderDomain, ctx.atsDomains),
      startsAt: null,
      endsAt: null,
      calendarEventId: null,
      applicationId: msg.thread_id ? (ctx.threadToApplication.get(msg.thread_id) ?? null) : null,
      location: null,
    };
  }

  return null;
}

function companyFromOrganiser(organiser: string | null, atsDomains: readonly string[]): string | null {
  const domain = organiser ? domainOfAddress(organiser) : null;
  if (!domain) return null;
  if (atsDomains.some((d) => matchesDomainSuffix(domain, d))) return null;
  const parts = domain.split('.');
  // "talent.nexi.it" → "Nexi". The resolver canonicalises this against the
  // company table anyway, but the fallback is sometimes the row that gets
  // created first — and a company called "nexi" in the interface looks like a
  // bug even when the matching is right.
  const label = parts.length >= 2 ? parts[parts.length - 2] : null;
  if (!label) return null;
  return label
    .split('-')
    .map((w) => (w ? w[0]!.toUpperCase() + w.slice(1) : w))
    .join(' ');
}

File: python/loop/ladder/rung2.py
"""Rung 2 — calendar and phrase heuristics.

"An .ics attachment or a meeting link from a company domain resolves the stage
by itself." 13% of messages, 2 ms, €0.

Two changes from the TypeScript, both recorded in `docs/port-to-python.md`:

  · Thread identity is no longer computed here. A reply on a known thread
    inherits its application with no parsing at all, which is an assertion about
    *identity* and not about intent — the ladder reads it once for every rung
    instead. In the TypeScript that assertion was dressed as an extraction with
    `intent: other`, and returning it early meant a rejection arriving on a
    known thread was never read by the vocabulary below.

  · An invitation whose title names no stage abstains instead of defaulting to
    `technical` (§3.2). A default wearing a reading's clothes is what produced a
    pipeline column of identical "Technical" stages.
"""

import re
from dataclasses import dataclass

from loop.domain import domain_of_address
from loop.domain.messages import CandidateMessage

from .company import company_from_display_name, company_from_domain
from .contracts import Extraction, LadderContext
from .domains import LEARNING_PLATFORMS, in_list
from .phrases import match_phrase

# The stage a calendar title implies.
#
# Interview invitations name their own stage more often than any other signal in
# the mailbox — "System & code review", "HR screening", "Final round" — and a
# title match is both cheaper and more reliable than asking a model.
_TITLE_TO_STAGE: tuple[tuple[re.Pattern[str], str], ...] = tuple(
    (re.compile(p, re.IGNORECASE), stage)
    for p, stage in (
        (r"\b(?:final|exec|leadership|founder|ceo|cto)\b", "final"),
        (r"\b(?:onsite|on-site|loop|super\s*day|assessment centre|full day)\b", "onsite_loop"),
        (r"\b(?:system\s*design|architecture|design (?:round|interview))\b", "system_design"),
        (
            r"\b(?:technical|coding|live\s*cod|pair\s*program|algorithm|code review)\b",
            "technical",
        ),
        (r"\b(?:take[- ]home|assignment|exercise|challenge)\b", "take_home"),
        (
            r"\b(?:hr|people|talent|recruiter|screening|intro|introductory|first call"
            r"|knowledge)\b",
            "hr_call",
        ),
    )
)

# "Every .ics invite from a company domain is a near-certain interview."
_INVITE_CONFIDENCE = 0.97
_CANCELLED_CONFIDENCE = 0.95
# One step below the same phrase from a known ATS: the sender is not
# established, so the claim is weaker even when the words are identical.
_PHRASE_CEILING = 0.88


def stage_from_title(title: str | None) -> str | None:
    if not title:
        return None
    return next((stage for pattern, stage in _TITLE_TO_STAGE if pattern.search(title)), None)


@dataclass(frozen=True, slots=True)
class HeuristicRung:
    costly: bool = False

    def extract(self, msg: CandidateMessage, ctx: LadderContext) -> Extraction | None:
        sender_domain = domain_of_address(msg.headers.sender)
        ats_domains = ctx.registry.ats_domains

        if msg.invite is not None:
            invite = msg.invite
            return Extraction(
                intent="interview_cancelled" if invite.cancelled else "interview_invite",
                confidence=_CANCELLED_CONFIDENCE if invite.cancelled else _INVITE_CONFIDENCE,
                rung=2,
                stage_hint=stage_from_title(invite.summary),
                company=company_from_domain(invite.organiser, ats_domains),
            )

        # An ATS sender has already been through rung 1, which tries the same
        # vocabulary; and a practice site sells using the words of hiring, so a
        # "coding challenge" promotion is not a take-home.
        if ctx.registry.is_ats(sender_domain) or in_list(sender_domain, LEARNING_PLATFORMS):
            return None

        phrase = match_phrase(f"{msg.headers.subject}\n{msg.text}")
        if phrase is None:
            return None
        return Extraction(
            intent=phrase.intent,
            confidence=min(phrase.confidence, _PHRASE_CEILING),
            rung=2,
            company=company_from_display_name(msg.headers.sender)
            or company_from_domain(sender_domain, ats_domains),
        )
